### ISSUE TITLE
{name: "SocketProtocolError", message: "Socket hung up", code: 1006}

### ISSUE THREAD

**Issue by mingkang1993** [zalmoxisus/remote-redux-devtools#48]
![e18fdff9-d084-45aa-bef3-4e2826840bdd](https://cloud.githubusercontent.com/assets/9126292/19297581/fef8a4a4-9078-11e6-8b0f-9d4b11114f8c.png)

**Comment by zalmoxisus** [comment]
Thanks for the report. Could you please check what version of `sc-broker-cluster` do you have installed? It seems related to https://github.com/SocketCluster/sc-broker-cluster/commit/ea5e7ba7ad5a46cd14ed1ff502120f5361af54fb, though I [don't see](https://github.com/SocketCluster/sc-broker-cluster/blob/master/index.js#L645) `stringify` function in use.

\cc @jondubois

**Comment by jondubois** [comment]
@zalmoxisus @mingkang1993 the dependency `sc-broker-cluster` should be `v2.1.6` or later. Maybe an earlier version was cached by npm from an earlier install? @mingkang1993 Try reinstalling/updating the dependency after doing `npm cache clean`.

**Comment by markortiz** [comment]
i've got a same error:
```
{
code:1006
message:"Socket hung up"
name:"SocketProtocolError"
...
}
```

my `sc-broker-cluster` is `v2.1.7`

**Comment by jondubois** [comment]
@markortiz Do you also get "...stringify is not a function" on the server side?
Are you running it over the network; if so, are you using plain HTTP/WS (not encrypted)?

Some corporate networks can block plain WebSocket traffic; the solution is to serve it over HTTPS.

Another thing that you can try is check that you're using the latest versions of https://www.npmjs.com/package/socketcluster and https://www.npmjs.com/package/socketcluster-server

Annother thing to check are your settings to make sure that it's connecting to the correct server URL.
This error just means that the server was lost or could not be found.

**Comment by markortiz** [comment]
@jondubois I'm just trying to run my react-native on local, here attached error screenshot:
![desctop screenshot](https://cloud.githubusercontent.com/assets/1829168/20257299/99cefe82-aa83-11e6-8338-cb47351ce29d.png)

Its still appear after updating `socketcluster`

**Comment by jondubois** [comment]
@markortiz Did you run `npm cache clean` and then reinstall remote-redux-devtools (and socketcluster)?
I cannot reproduce this - Maybe there is an old subdependency that is cached and not updating correctly.

**Comment by markortiz** [comment]
@jondubois I fix it by deleting `node_modules` and run `npm install` again and that error was gone. 
However it seems that I can't see any `state` nor `actions` on my **Redux Devtools extension on chrome** which I think a separate issue.

Thanks for the help! :)

**Comment by zalmoxisus** [comment]
@markortiz did you open Remote DevTools?
<img width="310" alt="redux_devtools" src="https://cloud.githubusercontent.com/assets/7957859/20259411/8b1937d4-aa5d-11e6-82c0-0e2e12eae635.png">

**Comment by markortiz** [comment]
@jondubois yes, I'm using `"remotedev-server": "^0.2.0"` and run `remotedev --hostname=localhost --port=8000 --injectserver=reactnative` then run `react-native run-android`

and I cant see any activities on **Redux Devtools**

![desctop screenshot2](https://cloud.githubusercontent.com/assets/1829168/20259866/20956272-aa92-11e6-9d0b-a7daa4011bf8.png)

**Comment by zalmoxisus** [comment]
Did you specified host and port on both the extension and `remote-redux-devtools` (and don't forget about `realtime` parameter)?

```js
  import { createStore, applyMiddleware } from 'redux';
  import { composeWithDevTools } from 'remote-redux-devtools';

  const composeEnhancers = composeWithDevTools({ realtime: true, port: 8000 });
  const store = createStore(reducer, /* preloadedState, */ composeEnhancers(
    applyMiddleware(...middleware),
    // other store enhancers if any
  ));
```

Just in case also [run `adb reverse`](https://github.com/zalmoxisus/remotedev-server#connect-from-android-device-or-emulator) (though it shouldn't be necessary when having `injectserver=reactnative`).

**Comment by markortiz** [comment]
@zalmoxisus thanks for your help! `adb reverse` was able to solve the problem. :)

**Comment by jakecraige** [comment]
I also ran into this when attempting to connect on an iOS device, and it was caused by a mismatch between the dev server's configuration and the store enhancer. I had:

```js
devTools({hostname: '10.0.0.4', port: 8000})
```

and

```sh
$ remotedev-debugger --hostname=localhost --port=8000 --injectserver=reactnative
```

Once I made the hostname of the shell command match, all worked :)

**Comment by darth-cheney** [comment]
I am experiencing the same WS error (`1006`, Connection Refused) when attempting to load the tools via a Django-hosted application:
  
```js
WebSocket connection to 'ws://localhost:8181/socketcluster/' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED
```
  
The relevant lines in my store look like this:
```js
import {composeWithDevTools} from 'remote-redux-devtools';
/* ... */
// DevTools setup
const composeEnhancers = composeWithDevTools({realtime: true, port: 8181});

const store = createStore(
    rootReducer,
    initState,
    composeEnhancers(applyMiddleware(thunk))
);
```
  
Per the above posts, I am using `sc-broker-cluster` v4.0.0.
  
None of this works even if I set the tools to use local hosting on any port or even when I attempt to match hostnames between the web server and the devtools.

**Comment by zalmoxisus** [comment]
@darth-cheney, check if your server is available on `http://localhost:8181`.

**Comment by darth-cheney** [comment]
I am using `atom-redux-devtools` and have attempted to match the hostname/port settings as in the store, but no dice. My understanding was that `atom-redux-devtools` took care of the server part -- is that not the case? Either way I cannot get this to work eve with the plain remotedev.io. 
  
I'm wondering now if this is at all even possible, since my webserver is Django and I cannot simply import a remotedev-server to start myself

**Comment by zalmoxisus** [comment]
@darth-cheney, if you don't use `remotedev.io` (which is by default), you should install and start [`remotedev-server`](https://github.com/zalmoxisus/remotedev-server):
```js
npm i -g remotedev-server && remotedev --port=8181
```
or if you don't want to install it globally:
```js
npm i -save-dev remotedev-server && ./node_modules/bin/remotedev --port=8181
```

It's not trivial to integrate it in an atom package, see https://github.com/zalmoxisus/atom-redux-devtools/issues/1.

**Comment by darth-cheney** [comment]
Ok I'll try to keep this as the last post, since it's veered off topic.
  
Initially, I tried the default (`remotedev.io`) and my store looked like this:
  
```js
import {composeWithDevTools} from 'remote-redux-devtools';
/* ... */
const store = createStore(
    rootReducer,
    initState,
    composeWithDevTools(applyMiddleware(thunk))
);
```
  
Additionally, I had `NODE_ENV` set to `development` (even though this isn't a node server).
  
However, nothing worked in either `atom-redux-devtools` or the `remotedev.io` development tools. Everything stayed blank and there were no warnings anywhere. I'd prefer _not_ to run a custom server if I don't have to, but went down that road when nothing was happening. Any other ideas?

**Comment by zalmoxisus** [comment]
Make sure to set `process.env.NODE_ENV`. For Webpack you should add it as following:
```js
plugins: [
  new webpack.DefinePlugin({
    'process.env.NODE_ENV': JSON.stringify('development')
  })
]
```

You can just set `realtime` to `true` or to other global variable to turn it off in production.

You might also want to try [our examples](https://github.com/zalmoxisus/remote-redux-devtools/tree/master/examples).

Feel free to provide more info, I'd happy to help.

**Comment by darth-cheney** [comment]
Yup, I'm an idiot. Needed to set `NODE_ENV` on the build for the JS bundle!
  
Thanks @zalmoxisus, this is working now. Much appreciated.

**Comment by zalmoxisus** [comment]
@darth-cheney no probs. I added the info to the README.

**Comment by SSTPIERRE2** [comment]
@zalmoxisus Man I've been at this for hours and cannot get this package to work with react-native run-android. As soon as I added `remotedev-debugger --hostname=localhost --port=8000 --injectserver=reactnative` as a postinstall script in my RN-redux app it cannot build. I get the error described above while building my app or even using the remotedev packages' `remotedev --hostname=localhost --port=8000` command directly. It says something about installing C++ compiler, do I need that? No docs say I need this.

**Comment by jondubois** [comment]
@SSTPIERRE2 Which version of Node.js are you using?
I think that this could be a combination of two issues:

1. Your machine doesn't have the latest C++ compiler to build the `uws` module as part of the regular installation process.
2. The version of `socketcluster-client` (`^5.3.1`) that is used doesn't have the pre-built binaries for `uws` for that specific Node.js version - New versions of `socketcluster-client` have the latest `uws` binaries so bumping up the dependency should help.

**Comment by SSTPIERRE2** [comment]
@jondubois Thanks for the quick reply!

I'm running Node 9.2, Windows 8.1, have Visual C++ 2015, just tried installing GCC, socketcluster-client ^9.0.0, and uws to no avail. I can't even use the remotedev command by itself, I just get get the C++ compile error.

**Comment by jondubois** [comment]
@SSTPIERRE2 It's likely related to the `uws` dependency on Windows specifically. I found this comment here that may help: https://github.com/nodejs/node-gyp/issues/1056#issuecomment-285131148

**Comment by SSTPIERRE2** [comment]
Hmm I still see the same uwebsockets compile error =/

**Comment by Dias999** [comment]
Hey guys.. I guess i'm a little late here, but just in case people have the same problem in the future.
I was having the same problem and I found out that my computer's clock was not in sync with the Android's clock.
So to fix it, I just had to set the time of the computer and the Android exactly the same (even the seconds).

**Comment by fabriziobertoglio1987** [comment]
I'm having this issue

**Comment by fabriziobertoglio1987** [comment]
> Hey guys.. I guess i'm a little late here, but just in case people have the same problem in the future.
> I was having the same problem and I found out that my computer's clock was not in sync with the Android's clock.
> So to fix it, I just had to set the time of the computer and the Android exactly the same (even the seconds).

```
Debugger and device times have drifted by more than 60s. Please correct this by running adb shell "date `date +%m%d%H%M%Y.%S`" on your debugger machine.
```
`adb shell "date `date +%m%d%H%M%Y.%S`"` did not fix the issue.
